python/tests/test_collections.py

```diff
@@ -0,0 +1,312 @@
+
+
+def test_nested_collection_not_found_returns_none() -> None:
+    gql = Mock(GQLClient)
+    _client = Client(gql)
+    gql.collection_create.return_value = _collection_create_collection_reference(
+        NESTED_COLLECTION_REFERENCE_KEY, NESTED_COLLECTION_REFERENCE_ID
+    )
+
+    result = collection(COLLECTION_NAME, _client)
+    gql.collection_create.return_value = _collection_create_collection_not_found(
+        NESTED_COLLECTION_ID
+    )
+
+    nested_result = result.collection(NESTED_COLLECTION_ID)
+
+    assert nested_result is None
+
+
+def test_collection_document_returns_new_document() -> None:
+    gql = Mock(GQLClient)
+    _client = Client(gql)
+    gql.collection_create.return_value = _collection_create_collection_reference(
+        COLLECTION_REFERENCE_KEY, COLLECTION_REFERENCE_ID
+    )
+    test_collection = collection(COLLECTION_NAME, _client)
+
+    document = test_collection.document(COLECTION_DOCUMNET_KEY)
+
+    assert isinstance(document, NumerousDocument)
+    assert document.exists is False
+
+
+def test_collection_document_returns_existing_document() -> None:
+    gql = Mock(GQLClient)
+    _client = Client(gql)
+    gql.collection_create.return_value = _collection_create_collection_reference(
+        COLLECTION_REFERENCE_KEY, COLLECTION_REFERENCE_ID
+    )
+    gql.collection_document.return_value = _collection_document_reference(
+        COLECTION_DOCUMNET_KEY
+    )
+    test_collection = collection(COLLECTION_NAME, _client)
+
+    document = test_collection.document(COLECTION_DOCUMNET_KEY)
+
+    assert isinstance(document, NumerousDocument)
+    assert document.exists
+
+
+def test_collection_document_set_data_uploads_document() -> None:
+    gql = Mock(GQLClient)
+    _client = Client(gql)
+    gql.collection_create.return_value = _collection_create_collection_reference(
+        COLLECTION_REFERENCE_KEY, COLLECTION_REFERENCE_ID
+    )
+    gql.collection_document_set.return_value = _collection_document_set_reference(
+        COLECTION_DOCUMNET_KEY
+    )
+    test_collection = collection(COLLECTION_NAME, _client)
+    document = test_collection.document(COLECTION_DOCUMNET_KEY)
+    assert isinstance(document, NumerousDocument)
+    assert document.exists is False
+
+    document.set({"test": "test"})
+
+    assert document.exists
+
+
+def test_collection_document_data_decoding_returns_dict() -> None:
+    gql = Mock(GQLClient)
+    _client = Client(gql)
+    gql.collection_create.return_value = _collection_create_collection_reference(
+        COLLECTION_REFERENCE_KEY, COLLECTION_REFERENCE_ID
+    )
+    gql.collection_document.return_value = _collection_document_reference(
+        COLECTION_DOCUMNET_KEY
+    )
+    test_collection = collection(COLLECTION_NAME, _client)
+    document = test_collection.document(COLECTION_DOCUMNET_KEY)
+
+    data = document.get()
+
+    assert isinstance(document, NumerousDocument)
+    assert document.exists
+    assert data
+    if data is not None:
```

Comment:
No conditionals in tests! :smile:  Assertions are what we use instead of conditionals. Why the `if` here?

Reply:
You also just `assert data` before, so the test will fail beforehand anyway. Basically the execution will always go into this branch

Reply:
Also assert the value instead of indexing. Then you wont get any index errors. `assert data == {"test": "test"}`

Also it is confusing when reading the tests why this is the expected assertion, because we only see the base64 decoded value in the test data. Can we make it easier to understand by perhaps encoding the test data inside the testfile?